metals/src/main/scala/scala/meta/internal/metals/concurrent/FileLock.scala

```diff
@@ -0,0 +1,38 @@
+package scala.meta.internal.metals.concurrent
+
+import scala.collection.concurrent.TrieMap
+
+import scala.meta.io.AbsolutePath
+
+/**
+ * Cache, where multiple files can write to the jar directory if
+ * we are not updating the meta file and deleting the old contents.
+ */
+object FileLock {
+
+  private val fileLocks =
+    new TrieMap[AbsolutePath, Boolean]()
+
+  def withWriteLock[A](dir: AbsolutePath)(f: () => A)(fallback: => A): A = {
+    val lock = fileLocks.get(dir)
```

Comment:
Not sure if this is fully safe: it seems that we could have both write and delete operations running at the same time, because the "check current lock state" and "write" are not an atomic transaction. Sorry if I have misunderstood the design or usage patterns of this code, but I think that `withWriteLock` would need to acquire the lock as well, and insert a value that other writers will understand as non-blocking. There are probably more problems with this to consider, e.g. how to ensure that the writer that originally acquired the lock won't release it while other writers are still live, maybe an atomic counter that the writers will increment and decrement? 🤔 

```scala
enum LockState
  case Free
  case Write(val counter: Long)
  case Delete
```

Reply:
There is a risk that we will be still writing when delete starts, but the risk is not too huge. I went with the different approach before and it became super compilcated :/

Reply:
Sure thing, I really have no idea how common or big of a problem it could be. 😅 In case an improvement would be needed maybe even `ReadWriteLock` from Java could help. 👍